### ISSUE TITLE
Erroneous error when code is unreachable

### ISSUE THREAD

**Issue by bcherny** [Microsoft/TypeScript#9616]
**TypeScript Version:**  2.0.0

**Code**

```ts
function infiniteLoop(): never {
  while (true) {
  }
  return "hello"
}
```

```sh
tsc test.js
```

**Expected behavior:**

```
test.ts(4,5): error TS7027: Unreachable code detected.
```

**Actual behavior:** 

```
test.ts(4,5): error TS7027: Unreachable code detected.
test.ts(4,12): error TS2322: Type 'string' is not assignable to type 'never'.
```

Because the `return` statement is unreachable, the compiler should not throw `TS2322`. `TS7027` is the real error, while `TS2322` is a red herring.

**Comment by mhegazy** [comment]
this looks right to me. there are two errors in this snippet. 1. the `while(true)` renders the return statement unreachable. and 2. the return statement type does not match the type annotation. removing the `while` does not remove the other error, so i do not see why it would be erroneous.

**Comment by bcherny** [comment]
The function will never (no pun intended) return a string, and the compiler knows that, so it seems overly conservative to complain about it. Maybe if it's to prevent potential errors the warning makes sense.
